unbound: added DNS64 support

danrl opened the pull request:
Maintainer: @EricLuehrsen 
Compile tested: X86_64, LEDE latest
Run tested: X86_64, LEDE latest

Description: Adds UCI support for DNS64

Signed-off-by: Dan Luedtke <mail@danrl.com>

danrl commented:
Supports https://github.com/openwrt/luci/pull/870

danrl commented on net/unbound/files/unbound.uci:
```diff
@@ -16,5 +16,7 @@ config unbound
 	option unbound_control '0'
 	option validator '0'
 	option validator_ntp '1'
-	#list domain_insecure ''
```
Ah, snap! That line should not be affected. Typo.

danrl replied:
Fixed!

danrl commented:
Thanks for the very clean source code @EricLuehrsen, was a pleasure working with it. I hope you like the changes.

EricLuehrsen commented:
(1) could you squash (s) or fixup (f) your corrective commit using interactive rebase?
(2) it seems you have DNS64 neccesarily tied to DNSSEC; is that intended?
- what if you string paste the module line in if-clauses then write `modules:`
- Unbound is fussy about module order and doesnt crash in repeatable way; i want to double check.

EricLuehrsen commented:
The split commits are making review notes difficult. Because these are fix-up commits, could you fixup "f" using interactive rebase? The following are comments I was trying to pin.
- It seems you have DNS64 neccesarily tied to DNSSEC; is that intended?
- DNSSEC or not-to-DNSSEC is at line 292 so this needs to merge in there somehow.
- What if we string-paste the module line on if-UNBOUND_B, then write the `module:` option after.
- Unbound is fussy about module order and doesnt crash in repeatable way; i want to double check.

hnyman commented:
Before this can get merged, you need to squash (rebase) the separate commits into one commit.
See advice at: https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md#advice-on-pull-requests

danrl commented:
@hnyman @EricLuehrsen Commits will be squashed of course before merging. It's a mess right now, but be assured I will squash everything before we merge. Maybe in a new, cleaner PR. 

But let's sort out the concerns first.

> * It seems you have DNS64 neccesarily tied to DNSSEC; is that intended?
> * DNSSEC or not-to-DNSSEC is at line 292 so this needs to merge in there somehow.

To be honest, I can't see that. DNSSEC starts in 292 but ends somewhere at 314. The code I added starts later in the file, somewhere at 429 IIRC. May I ask you to explain where the dependency comes from. It was not intended and I think there is no good reason for that dependency, is there?

> * What if we string-paste the module line on if-UNBOUND_B, then write the module: option after

Sorry, not sure what this means. Doesn't UNBOUND_B stand for a `boolean` typed variable? Have I missed something?

> Unbound is fussy about module order and doesnt crash in repeatable way; i want to double check.

I have been running quite a few systems where the DNS64 module was the last module in the config. Admittedly, I never tried any other order. The order in this PR runs fine on my test system (X86_64 in VirtualBox, built from latest LEDE sources on a daily basis). Did it crash on other platforms?

> UCI address dns54_prefix should be dns64_prefix

Done!
Also, in an earlier comment you mentioned we may not use the well-known DNS64 `64:ff9b::/96` in UCI files, but rather use the well-known prefix in case the user did not specify any prefix. I changed that as requested and think it is the better option.

EricLuehrsen commented:
With the commit noise it makes using github web review tool difficult, and we dont know if we are taging active material. So I will do my best to help, but out of context, some short hand comments may be confusing.
- Unbound module "validator" is DNSSEC and your "dns64" line includes it.
- you can see the module string written in the if-else chain at 292 
- from 292 on... we need a clever way to mix n match options.
- Unbound can be inconsistent with the wrong module order
- Unbound will not always crash, but maybe dnssec disables

EricLuehrsen commented:
- I have created a branch in my github repo with a more robust solution to what I think you want
- Do not remove your 'dns64' branch or this pull request dies
- Copy your branch to 'dns64_scratch' branch to keep your history
- Rebase your 'dns64' branch to drop your commits
- Rebase your 'dns64' on openwrt/master latest
- Cherry-pick the commits I give you
    - I have each file as one commit (presquash) so fix-ups are easier to manage
    - You will be attributed as "danrl commit with EricLuehrsen" from SoB native merging
- Test, edit, squash, and push to your 'dns64' on github
```
$ cd <your project dir>/openwrt-packages/
$ git remote add eric https://github.com/EricLuehrsen/openwrt-packages
$ git remote -v
  <should see upstream, origin, eric repo remote links>
$ git fetch --all
$ git checkout dns64
$ git rebase upstream/master
$ git push -f
$ git branch dns64_scratch
$ git push -u origin dns64_scratch
  <your current work is saved away locally and on github; check website>
$ git rebase -i HEAD~10
  <drop "d" your commits to "unbound:">
$ git cherry-pick 28a6cc8
$ git cherry-pick 98d3111
$ git cherry-pick 1497eff
$ git cherry-pick 12deca3
  <these will be found in your-local-copy of my-repo and applied>
$ git rebase upstream/master
$ git push -f 
  <you should now be 4 commits ahead of master on dns64; check website>
```
The commit messages should automatically update that we worked together on this. You can now test and make modifications if it doesn't work as you like. Then squash all of these into one commit. This pull request will continue to track your work. Let me know how it goes.

danrl commented:
Thanks very much @EricLuehrsen, i will take care of this tomorrow. Sorry, was on the road the whole day.

danrl commented:
Done! I understand now what your concerns regarding the module string were. I'll test the code on a build from today's sources and will post results.

danrl commented:
Tested and it works like a charm. 👍

EricLuehrsen commented:
Glad it works. When you are happy with enough "seat time", then rebase-squash and rebase-reword.

danrl commented:
I think we are all set.

EricLuehrsen commented:
Ready to merge.
